In [1]:
import pandas as pd
import numpy as np
import sys
import re

import seaborn as sns
import matplotlib.pyplot as plt

In [2]:
sys.path.append('../')
import src.limpieza as lmp

In [3]:
import requests

In [4]:
pd.options.display.max_columns = None

In [5]:
url = 'https://www.ine.es/jaxiT3/files/t/es/csv_bdsc/2078.csv'
res1 = requests.get(url)


In [6]:
res1.status_code

200

In [7]:
from pathlib import Path

data_folder = Path(r"C:\Users\mituc\Ironhack\Curso\IronLabs\Proyecto-1\data")
file_to_open = data_folder / "2078.csv"
with open(file_to_open, 'wb') as f:
    f.write(res1.content)

In [8]:
df_guiris = pd.read_csv(r"C:\Users\mituc\Ironhack\Curso\IronLabs\Proyecto-1\data\2078.csv", sep=';', encoding='latin-1', index_col = 0)

In [9]:
df_guiris.shape

(86920, 4)

In [10]:
df_guiris.head(5)

Viajeros y pernoctaciones             Residencia  \
Puntos turísticos                                                    
01059 Vitoria-Gasteiz                   Viajero  Residentes en España   
01059 Vitoria-Gasteiz                   Viajero  Residentes en España   
01059 Vitoria-Gasteiz                   Viajero  Residentes en España   
01059 Vitoria-Gasteiz                   Viajero  Residentes en España   
01059 Vitoria-Gasteiz                   Viajero  Residentes en España   

                       Periodo   Total  
Puntos turísticos                   
01059 Vitoria-Gasteiz  2022M01  11.759  
01059 Vitoria-Gasteiz  2021M12  16.496  
01059 Vitoria-Gasteiz  2021M11  16.667  
01059 Vitoria-Gasteiz  2021M10  21.467  
01059 Vitoria-Gasteiz  2021M09  18.647

In [11]:
df_guiris.duplicated().sum()

12322

In [12]:
df_guiris[df_guiris.duplicated() == True]

Viajeros y pernoctaciones                   Residencia  \
Puntos turísticos                                                          
02003 Albacete                          Viajero        Residentes en España   
02003 Albacete                          Viajero        Residentes en España   
02003 Albacete                          Viajero        Residentes en España   
02003 Albacete                          Viajero  Residentes en el Extranjero   
02003 Albacete                          Viajero  Residentes en el Extranjero   
...                                         ...                          ...   
50297 Zaragoza                   Pernoctaciones        Residentes en España   
50297 Zaragoza                   Pernoctaciones        Residentes en España   
50297 Zaragoza                   Pernoctaciones  Residentes en el Extranjero   
50297 Zaragoza                   Pernoctaciones  Residentes en el Extranjero   
50297 Zaragoza                   Pernoctaciones  Residentes en el Extranjero   

                       Periodo Total  
Puntos turísticos                 
02003 Albacete         2020M06     .  
02003 Albacete         2020M05     .  
02003 Albacete         2020M04     0  
02003 Albacete         2020M06     .  
02003 Albacete         2020M05     .  
...                        ...   ...  
50297 Zaragoza         2020M05     .  
50297 Zaragoza         2020M04     0  
50297 Zaragoza         2020M06     .  
50297 Zaragoza         2020M05     .  
50297 Zaragoza         2020M04     0  

[12322 rows x 4 columns]

In [13]:
df_guiris[df_guiris['Total'] == '0']

Viajeros y pernoctaciones                   Residencia  \
Puntos turísticos                                                          
01059 Vitoria-Gasteiz                   Viajero        Residentes en España   
01059 Vitoria-Gasteiz                   Viajero  Residentes en el Extranjero   
01059 Vitoria-Gasteiz            Pernoctaciones        Residentes en España   
01059 Vitoria-Gasteiz            Pernoctaciones  Residentes en el Extranjero   
02003 Albacete                          Viajero        Residentes en España   
...                                         ...                          ...   
49275 Zamora                     Pernoctaciones  Residentes en el Extranjero   
50297 Zaragoza                          Viajero        Residentes en España   
50297 Zaragoza                          Viajero  Residentes en el Extranjero   
50297 Zaragoza                   Pernoctaciones        Residentes en España   
50297 Zaragoza                   Pernoctaciones  Residentes en el Extranjero   

                       Periodo Total  
Puntos turísticos                 
01059 Vitoria-Gasteiz  2020M04     0  
01059 Vitoria-Gasteiz  2020M04     0  
01059 Vitoria-Gasteiz  2020M04     0  
01059 Vitoria-Gasteiz  2020M04     0  
02003 Albacete         2020M04     0  
...                        ...   ...  
49275 Zamora           2020M04     0  
50297 Zaragoza         2020M04     0  
50297 Zaragoza         2020M04     0  
50297 Zaragoza         2020M04     0  
50297 Zaragoza         2020M04     0  

[425 rows x 4 columns]

Aunque parecía que había duplicados, en la tabla anterior podemos ver que no es así y si resetamos el index desaparecen

In [14]:
df_guiris.reset_index(inplace=True)

In [15]:
df_guiris.duplicated().sum()

0

In [16]:
df_guiris.dtypes

Puntos turísticos        object
Viajeros y pernoctaciones    object
Residencia                   object
Periodo                      object
Total                        object
dtype: object

Tendremos que limpiar algunos nombres

In [17]:
df_guiris.columns

Index(['Puntos turísticos', 'Viajeros y pernoctaciones', 'Residencia',
       'Periodo', 'Total'],
      dtype='object')

In [18]:
print(df_guiris['Viajeros y pernoctaciones'].unique())
print(df_guiris['Residencia'].unique())

['Viajero' 'Pernoctaciones']
['Residentes en España' 'Residentes en el Extranjero']


In [19]:
df_guiris.isnull().sum()

Puntos turísticos           0
Viajeros y pernoctaciones       0
Residencia                      0
Periodo                         0
Total                        5358
dtype: int64

In [20]:
df_guiris.shape

(86920, 5)

Renombramos las columnas

In [21]:
df_guiris = df_guiris.rename(columns={'Puntos turísticos': 'Puntos_turisticos'})
columnas_limpias = {col : col.lower().replace(" ", "_") for col in df_guiris.columns}
df_guiris.rename(columns = columnas_limpias, inplace = True)

In [22]:
df_guiris.columns

Index(['puntos_turisticos', 'viajeros_y_pernoctaciones', 'residencia',
       'periodo', 'total'],
      dtype='object')

Ahora vamos a dejar limpita la columna 'total'. Primero retiraremos los puntos y luego convertiremos los valores en enteros.

Es importante tener en cuenta que en este csv si aparece un punto '.' a secas, significa bien que el dato está protegido por la ley de secreto estadístico, (Artículo 12 de la Ley 7/2006) o en algunas ocasiones que el dato no se recogió. y Para nosotros es como si fuera un nulo.

Esta operación de limpieza tiene por tanto una doble función:
- poder convertir los string en int
- convertir los '.' en NaN para manejarlos después

In [23]:
df_guiris['total'] = df_guiris['total'].str.replace('.', '')

C:\Users\mituc\AppData\Local\Temp/ipykernel_14436/2279084469.py:1: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  df_guiris['total'] = df_guiris['total'].str.replace('.', '')


In [24]:
df_guiris['total'] = pd.to_numeric(df_guiris['total'], errors='coerce').astype('Int64')

Podemos comprobar que los nulos han aumentado de 5358 a 13056

In [25]:
df_guiris.isnull().sum()

puntos_turisticos                0
viajeros_y_pernoctaciones        0
residencia                       0
periodo                          0
total                        13056
dtype: int64

Sabemos por la documentación del csv que entre abril y junio del 2020 no hubo recogida de datos a causa de la crisis del covid-19. Veamos cómo están esos datos.

In [26]:
df_guiris[(df_guiris['periodo'] == '2020M04') | (df_guiris['periodo'] == '2020M05') | (df_guiris['periodo'] == '2020M06')]

puntos_turisticos viajeros_y_pernoctaciones  \
19     01059 Vitoria-Gasteiz                   Viajero   
20     01059 Vitoria-Gasteiz                   Viajero   
21     01059 Vitoria-Gasteiz                   Viajero   
224    01059 Vitoria-Gasteiz                   Viajero   
225    01059 Vitoria-Gasteiz                   Viajero   
...                      ...                       ...   
86530         50297 Zaragoza            Pernoctaciones   
86531         50297 Zaragoza            Pernoctaciones   
86734         50297 Zaragoza            Pernoctaciones   
86735         50297 Zaragoza            Pernoctaciones   
86736         50297 Zaragoza            Pernoctaciones   

                        residencia  periodo  total  
19           Residentes en España  2020M06   <NA>  
20           Residentes en España  2020M05   <NA>  
21           Residentes en España  2020M04      0  
224    Residentes en el Extranjero  2020M06   <NA>  
225    Residentes en el Extranjero  2020M05   <NA>  
...                            ...      ...    ...  
86530        Residentes en España  2020M05   <NA>  
86531        Residentes en España  2020M04      0  
86734  Residentes en el Extranjero  2020M06   <NA>  
86735  Residentes en el Extranjero  2020M05   <NA>  
86736  Residentes en el Extranjero  2020M04      0  

[1272 rows x 5 columns]

Dado que en los meses del confinamiento toda la actividad turística estuvo suspendida, vamos a sustituir los nulos por 0, pues no es tanto que no hubiese recogida de datos como que no había turistas.

In [27]:
# # hacer diccionario o lista y sustituir con condicional
# df_guiris.loc[(df_guiris['periodo'] == '2020M04') | 
#               (df_guiris['periodo'] == '2020M05') | 
#               (df_guiris['periodo'] == '2020M06'), 'total'] = 0

In [28]:
covid = ['2020M04', '2020M05', '2020M06']

df_guiris.loc[(df_guiris['periodo'].isin(covid)), 'total'] = 0

In [29]:
# covid = ['2020M04', '2020M05', '2020M06']

# for confinamiento in covid:    
#     for fecha in df_guiris['periodo']:
#         if fecha == confinamiento:
#             df_guiris.loc[df_guiris['periodo'] == fecha, 'total'] = 0

Y comprobamos que los nulos han descendido de 13056 a 12208

In [30]:
df_guiris.isnull().sum()

puntos_turisticos                0
viajeros_y_pernoctaciones        0
residencia                       0
periodo                          0
total                        12208
dtype: int64

Los nulos que quedan se deben a que o bien no había recogida de datos en esa fecha para esa zona turística, o bien el dato no está publicado porque está protegido por secreto estadístico.
En todo caso no se pueden eliminar alegremente. En el caso de que falten en meses puntuales pero dispongamos de datos para el resto del año, quizá convendría sustiruirlos.

In [31]:
null_rows = pd.isnull(df_guiris['total'])

In [32]:
df_guiris[null_rows].to_csv('nulls.csv')

In [33]:
df_guiris['periodo'] = df_guiris['periodo'].str.replace('M', '-')

In [34]:
df_guiris.to_pickle('guiris.pkl')
#df_guiris = pd.read_pickle('guiris.pkl')

In [35]:
df_guiris['periodo'] = pd.to_datetime(df_guiris['periodo'], format='%Y%m%')
df_guiris['periodo'] = df_guiris['periodo'].dt.date.apply(lambda x: x.strftime('%Y-%m'))

In [36]:
df_guiris.dtypes

puntos_turisticos            object
viajeros_y_pernoctaciones    object
residencia                   object
periodo                      object
total                         Int64
dtype: object

In [37]:
df_guiris.set_index('periodo', inplace=True)

Renombramos los valores

In [38]:
df_guiris['residencia'].replace({'Residentes en España': "residentes_espana", 'Residentes en el Extranjero': 'residentes_en_el_extranjero'}, inplace=True)
df_guiris['viajeros_y_pernoctaciones'] = df_guiris['viajeros_y_pernoctaciones'].str.lower()

Aplicamos la función que limpia y organiza los datos por ciudades

In [39]:
lista = ['Barcelona']

In [40]:
lmp.pastor_de_guiris(df_guiris, lista)

In [41]:
df_guiris = pd.read_pickle('Barcelona_guiris_esquilados.pkl')

In [42]:
df_guiris.head(5)

viajeros_espana  viajeros_extranjero  pernoctaciones_espana  \
periodo                                                                
2022-01           107162               194312                 203683   
2021-12           144299               250133                 282398   
2021-11           120748               314677                 228675   
2021-10           130404               364349                 270494   
2021-09           102053               309572                 207329   

         pernoctaciones_extranjero  media_pernoc_esp  media_pernoc_ext  \
periodo                                                                  
2022-01                     541541               1.9               2.8   
2021-12                     674630               2.0               2.7   
2021-11                     871021               1.9               2.8   
2021-10                    1013514               2.1               2.8   
2021-09                     856562               2.0               2.8   

         total_viajeros  total_pernoctaciones  media_total_pernoc  
periodo                                                            
2022-01          301474                745224                 2.5  
2021-12          394432                957028                 2.4  
2021-11          435425               1099696                 2.5  
2021-10          494753               1284008                 2.6  
2021-09          411625               1063891                 2.6